### DISCUSSION TITLE
Refactor widgets + third-party widgets

### DISCUSSION THREAD

**Discussion by danilopolani** [filamentphp/filament#217] [Ideas]
Hello,
is there a way currently to create a 3rd party widget? I found out the `Filament::registerWidget` method but unfortunately it will look through the `App/Widgets` folder and other paths defined in the configuration (I guess?), so I got no lucky with that.

Maybe would be easier to manage widgets like this? With this configuration users would be able to remove undesired widgets (without a `disable` flag), re-order them easily and create a more flexible layout.

**`config/filament.php`**

```php
'widgets' => [
	// Auto row (2 cols)
	[
		\Filament\Widgets\Welcome::class,
		\Filament\Widgets\Info::class,
	],

	// Full row
	\ThirdParty\Widgets\ThirdPartyWidget::class,
],
```

What do you think? If that works to you I could try to draft a PR, although I don't know how much time I have to work on it

**Comment by ryangjchandler** [comment]
@danilopolani `registerWidget` will register any custom widgets outside of the `app/Filament/Widgets` directory, you need to call it inside of a Service Provider.

**Reply by danilopolani** [reply]
Got it finally, I had to register the component to Livewire too:

**PackageServiceProvider.php**
```php
use Package\Http\Livewire\MyWidget;

public function boot()
{
        Livewire::component('my-widget', MyWidget::class);
        Filament::registerWidget(MyWidget::class);
}
```


I don't know if we're supposed to register the livewire component too, but I think so, because without that it just returns the error `Unable to find component: []` from Livewire (basically can't find the alias? Also after a `php artisan livewire:discover`)

**Reply by ryangjchandler** [reply]
@danilopolani I'll open a pull request tomorrow to ease this process. I'd also recommend using the `PluginServiceProvider` as your base class (extend it) since it has some helpers for registering things.

```php
class PackageServiceProvider extends PluginServiceProvider
{
    protected $widgets = [
        MyWidget::class,
    ];
}
```

I'll open a PR so that you also don't need to manually register your Livewire component.

**Reply by danilopolani** [reply]
That's huge, thank you! For the quick answers too on Sunday haha.
